### ISSUE TITLE
Position of tooltip is of on transformed elements

### ISSUE THREAD

**Issue by Daanoz** [ajaxorg/ace#2629]
The position of the tooltip is off, when the editor has a parent that uses the css transform property to position itself. For example modal dialogs use this to center the content.

A codepen to illustrate the problem: http://codepen.io/anon/pen/doLPOP

The transform property creates it's own coordinate system, causing 'position:fixed' to be relative to the container instead of the page.
A solution would be to add to tooltip to the body element of the page instead of contained within the editor element or to calculate a relative offset to the editor instead of using the events clientX / clientY.

**Comment by Neill83** [comment]
can someone check this and fix it please? @nightwing ?

**Comment by tsmirnov** [comment]
It seems, that it's pretty hard issue. Nobody have fixed it for 2 years. What if we just add config option, like "disableTooltip"? I think it is easy and quick solution. I can work on PR. What do you think?

**Comment by mczepiel** [comment]
I notice  the autocomplete "tooltips" don't exhibit this behavior as they are presented in a DOM node outside of the editor's container.

In addition to not being subjected to any transformations on that container, this also means they appear in a different stacking context, which may or may not be desirable depending on your scenario.

Anyway, there's no easy way to influence the `MouseHandler` or override how it creates a `GutterHandler`, or how that creates a `GutterTooltip`.

Ideally, you could inject your own objects somewhere in that chain and re-root your tooltips.

As a quick hack, you _can_ alter the base `Tooltip.prototype.$init` and set `this.$parentNode` right before the tooltip creates its own node and appends it to said container.

e.g.
```javascript
const init = ToolTip.prototype.$init;
const tooltipContainer = document.body; // or something;

Tooltip.prototype.$init = function () {
  // if (constructor.name === 'GutterTooltip') or however you want to decide this
  this.$parentNode = tooltipContainer;
  init.call(this);
}
```

Not a great solution, but possibly a way to see if it's worth pursuing work to make it easier to influence the parentContainer that's hard-coded right now.

**Comment by EmaGht** [comment]
Have the same issue here... is there something Work in Progress or a workaround to prevent this?
Thanks

**Comment by blongstreth** [comment]
So, I ran into this very issue using Angular.js UI bootstap modal.  I tried various things and found the workaround proposed by @mczepiel works pretty well, so thank you!.  Also, I noticed that the AcePopup component is not placed in the editor container, but rather the body or document element (similar to this workaround).  

Example usage in ext-language_tools.js:
`this.popup = new AcePopup(document.body || document.documentElement);`

I have found that the AcePopup keeps creating elements on the body creating a memory leak since it seems there is no element cleanup.  This is kind of an issue when the editor is used in a modal.  I probably will file an issue for it if it doesn't already exist.
